### PR TITLE
[Navigation] Gracefully handle deprecation of classes

### PR DESCRIPTION
**Pull request by psealock**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds a `RemovedDeprecated` class to handle removed Navigation classes. This way, method calls to these classes will not create fatal errors. Instead, use logger and Tracks to alert developers that these classes should be removed.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Using the Snippets plugin add a new Snippet with the following code to simulate a problematic extension trying to call a removed class.
```php
use Automattic\WooCommerce\Admin\Features\Navigation\Menu;

if ( ! function_exists( 'testFatalCausingImpletmentation' ) ) {
	function testFatalCausingImpletmentation() {
		// call static method
		Menu::add_plugin_item();

		// call instance method
		$menu = new Menu();
		$menu->add_plugin_item();
	}
}

testFatalCausingImpletmentation();
```
2. Navigate to WooCommerce > Home. See that no fatals occur.
3. WooCommerce > Status > Logs > plugin-woocommerce. See the logged out message relating to the misuse of removed classes.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Gracefully handle deprecation of Navigation feature classes.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php]: If I understand correctly, I think the `get_called_class()` is intended for use in static methods to get the class name that the static method is called on. In the context of an instance method like `__call`, it should be `get_class($this)` to correctly retrieve the class name of the current instance, thought I think it doesn't matter in this case.

Comment (chihsuan) [reply plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php]: The log message looks misleading. It prints the alias name of the class, not the original class name. But it's unrelated to the `get_called_class()` issue mentioned above.


Screenshot 2024-10-23 at 14 41 06

Comment (chihsuan) [reply plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php]: This approach could log the original class name instead of the alias name.

```php
class RemovedDeprecated {
	/**
	 * The original class name for an alias class.
	 *
	 * @var string|null
	 */
	private static $originalClass;

	/**
	 * Set the original class name for an alias class.
	 *
	 * @param string $originalClass The original class name.
	 */
	public static function __constructStatic( $originalClass ) {
		self::$originalClass = $originalClass;
	}

	private static function handle_deprecated_method_call( $class, $name ) {
		$logger = wc_get_logger();

		if ( $logger ) {
			$logger->warning(
				self::$originalClass . " is deprecated since 9.3 with no alternative. Please remove the call to $name."
			);
		}

		if ( class_exists( 'WC_Tracks' ) ) {
			WC_Tracks::record_event( 'deprecated_navigation_method_called' );
		}
	}
```

```php
// When creating aliases
RemovedDeprecated::__constructStatic( \Automattic\WooCommerce\Admin\Features\Navigation\Menu::class );
class_alias( RemovedDeprecated::class, \Automattic\WooCommerce\Admin\Features\Navigation\Menu::class );

RemovedDeprecated::__constructStatic( \Automattic\WooCommerce\Admin\Features\Navigation\Screen::class );
class_alias( RemovedDeprecated::class, \Automattic\WooCommerce\Admin\Features\Navigation\Screen::class );

RemovedDeprecated::__constructStatic( \Automattic\WooCommerce\Admin\Features\Navigation\CoreMenu::class );
class_alias( RemovedDeprecated::class, \Automattic\WooCommerce\Admin\Features\Navigation\CoreMenu::class );

```

**Review by moon0326 — COMMENTED** [review]
Other than what Chi already mentioned, it looks good to me! I'll let Chi approve it when he's ready.

File: plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php
Comment (psealock) [reply plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php]: >The log message looks misleading.

Good point!

>This approach could log the original class name instead of the alias name.

Indeed this works, but `$originalClass` gets overwritten each time because there isn't a new instance. I've tried mapping, but thats not working with all the classes getting aliased to the same `RemovedDeprecated`. I'm not quite seeing a way other than to create 3 separate `RemovedDeprecated` classes.

I could simply log "Navigation classes are deprecated since 9.3 with no alternative..."

What do you think?

Comment (chihsuan) [reply plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php]: I think `Navigation classes` is still a bit vague. It would be better to mention the specific namespace or class name.

How about `Navigation classes Automattic\WooCommerce\Admin\Features\Navigation\* are deprecated since 9.3 with no alternative...` 

or `The WooCommerce Admin Navigation feature and its classes (Screen, Menu, CoreMenu) are deprecated since 9.3 with no alternative...`?

Comment (psealock) [reply plugins/woocommerce/src/Admin/Features/Navigation/RemovedDeprecated.php]: Thanks @chihsuan, fixed in https://github.com/woocommerce/woocommerce/pull/52248/commits/10a2a3d222b050430fdbc0263d3632e0625864d2

**Review by chihsuan — APPROVED** [review]
Thanks for addressing feedback, LGTM!

**Pull request merged**
